Implement C++ functionality to post-process profiler data

sagarwalTT opened the pull request:
#14846 
#27079 

This PR adds support for profiler data post-processing in C++. It can be enabled with `TT_METAL_PROFILER_CPP_POST_PROCESS=1`, by default it is disabled.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/18862052930) ()
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (https://github.com/tenstorrent/tt-metal/actions/runs/18883984068/attempts/1)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/18862056400)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/18862061639)
- [x] [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/18862066523)
- [x] [T3K Profiler] (https://github.com/tenstorrent/tt-metal/actions/runs/18884002702)
- [x] [Metal Microbenchmarks] (https://github.com/tenstorrent/tt-metal/actions/runs/18862075684)
- [x] New/Existing tests provide coverage for changes

sagarwalTT committed:
Saving

sagarwalTT committed:
Saving

sagarwalTT committed:
Saving

sagarwalTT committed:
Saving

sagarwalTT committed:
Saving

sagarwalTT committed:
Saving

sagarwalTT committed:
Saving

sagarwalTT committed:
Cleanup

sagarwalTT committed:
Saving

sagarwalTT committed:
Modify compare_ops_logs script and how it's called from run_profiler_regressions

sagarwalTT committed:
Change CORE_AGG to TENSIX_RISC_AGG

sagarwalTT committed:
Change CORE_AGG to TENSIX_RISC_AGG

sagarwalTT committed:
Add functionality to load analysis configs from JSON file

sagarwalTT committed:
Fix rebase conflicts

sagarwalTT committed:
Saving

sagarwalTT committed:
Saving

sagarwalTT committed:
Cleanup

sagarwalTT committed:
Saving

sagarwalTT committed:
Saving

sagarwalTT committed:
Saving

sagarwalTT committed:
Fix test called in T3K script

sagarwalTT committed:
Rename firmware duration analysis name

sagarwalTT committed:
Fix compare script

sagarwalTT committed:
Update common columns num assert

sagarwalTT committed:
Remove call to script to compare Python device-only ops perf report with C++ ops perf report

sagarwalTT commented:
/codeowners

sagarwalTT committed:
Add C++ post-proc test

sagarwalTT committed:
Fix pre-commit errors

bbradelTT commented on tt_metal/impl/profiler/profiler_analysis.cpp:
```diff
@@ -0,0 +1,369 @@
+
+        auto& it = op_id_to_meta_data[{marker.runtime_host_id, marker.trace_id, marker.trace_id_counter}];
+
+        TT_ASSERT(it.device_id == marker.chip_id);
```
Please use TT_FATALs. TT_ASSERT would only work in Debug builds.

sagarwalTT committed:
Clean up comparison script

mo-tenstorrent commented on tests/ttnn/tracy/test_perf_op_report.py:
```diff
@@ -104,3 +119,22 @@ def test_tensor_io(self, run_test_do_post_proc):
+        request = run_test_do_cpp_post_proc
+        python_ops_perf_report = get_latest_ops_log_filename(request.param["name"])
+        cpp_ops_perf_report = (
+            generate_logs_folder(get_profiler_folder(request.param["name"])) / PROFILER_CPP_DEVICE_OPS_PERF_REPORT
```
Do we need the paths to be passed?

mo-tenstorrent reviewed (approved):
Looks great!

Left a minor comment on the test script

Please run BH post commit, T3K profiler and Device perf report CIs

sagarwalTT committed:
Fix issue with profiler.cpp not detecting path to cpp_device_analyses.json

sagarwalTT committed:
Cleanup

sagarwalTT commented:
/codeowners

sagarwalTT committed:
Use variable for C++ ops perf report name

sagarwalTT committed:
Update tracy submodule

sagarwalTT committed:
Fix conflict after rebasing

sagarwalTT committed:
Fix issues after rebasing

jbaumanTT reviewed (approved):
Since you added some asserts, make sure you try running All Post Commit using a RelWithDebInfo build type.

sagarwalTT committed:
Initialize primitive types in profiler analysis structs

sagarwalTT committed:
Enable mid-run C++ post-processing for the profiler (#31272)

[x] [T3K Profiler]
(https://github.com/tenstorrent/tt-metal/actions/runs/18860968983)

---------

Co-authored-by: mo-tenstorrent <mmemarian@tenstorrent.com>

sagarwalTT commented:
> Will this make our device perf pipeline run faster?

No, this change won't, but I'm working on another change after this PR gets merged in that should make them run faster.

sagarwalTT committed:
Account for chip frequency when calculating duration

tt-rkim reviewed (commented):
We're not allowing for increases in runtime on APC

Please look into what's going on with your test runtimes

sagarwalTT committed:
Fix clang-tidy errors
